Fix transfer results transaction aborted

Hagellach37 opened the pull request:
This refers to #478 

Here we try to improve the transfer of results from firebase to postgres. 

Basically, we keep the existing logic, but don't use a Firebase transaction function anymore. 

* [x] check for potential exceptions and handle

Matthias-Schaub commented on mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py:
```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
+        # and then delete these results from Firebase.
+        # In case something goes wrong during the insert, results in Firebase
+        # will not get deleted.
+        delete_results_from_firebase(project_id, results)
```
Here we have to be careful. This should only be executed if `save_results_to_postgres` has been successful. Currently we hope that this function (and the psycopg library) will raise an exception when inserting results to the db has not been successful. I do not know if there is need for other checks and how they would look like.

laurentS replied:
To second what Matthias said, you want to make sure that you delete results only if there is no error when saving to postgres. You could use the python exception construct:
```python
try:
    save_results_to_postgres()
except SomePGError:
    log("failed")
else:
    delete_results_from_firebase()  # this is only run if there is no exception in the try block
```
See https://docs.python.org/3/tutorial/errors.html for more details. This only works if any problem in the postgres execution raises an exception, but I think this is a fairly safe assumption (reading the docs never hurts ;)

Hagellach37 replied:
Hey, 
yep this is a very good comment. I changed it based on your comments.

laurentS replied:
Hmm, just to be on the safe side, would it make sense to add `except Exception:` before the `else:` block here? The way you do it now, if any exception that is not a `ForeignKeyViolation` is raised in the `try` block, it won't get handled, and the process will crash. If you add another handler, you can get some cleaner logging, and keep going. I'm not sure what makes more sense.

Hagellach37 replied:
yes, you are right. I added some new lines.

Matthias-Schaub commented on mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py:
```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
+    for group_id, users in results.items():
+        for user_id, result in users.items():
+            key = f"{group_id}/{user_id}"
+            data[key] = None
```
Does setting the reference (`fb_db.reference(f"v2/results/{project_id}/")`) with a dict containing this key (`"{group_id}/{user_id}"`) work as expected?

I would build a nested dict as follows:
```python
    for group_id, users in results.items():
        data[group_id] = {}
        for user_id in users.keys():        
             data[group_id][user_is] = None
```

But it is a while since I worked with Firebase.

Hagellach37 replied:
Hey, @Matthias-Schaub 
it works. So basically the first part in that dict is the path. That's why it looks a bit strange.

This is also covered by the tests, in which we check that the results are actually deleted from firebase.

laurentS commented on mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py:
```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
+    uploading results to Firebase at the same time. Basically, this is
+    due to the behaviour of Firebase Transaction function:
+
+        "If another client writes to this location
```
Can I suggest adding a link to the source? This page is not what you quoted, but also has helpful info on transactions https://firebase.google.com/docs/firestore/manage-data/transactions#transaction_failure

Hagellach37 replied:
I added a link to the Firebase python sdk where I copied the statement from.

You link refers to FireStore, which is another thing. :)

laurentS commented on mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py:
```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
+        # In case something goes wrong during the insert, results in Firebase
+        # will not get deleted.
+        delete_results_from_firebase(project_id, results)
+        del fb_db
```
Any special reason you delete `fb_db` here? it will get deleted automatically next line, when it becomes out of scope.

Hagellach37 replied:
no, removed

laurentS commented on mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py:
```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
+            data[key] = None
 
+    results_ref = fb_db.reference(f"v2/results/{project_id}/")
+    results_ref.update(data)
```
What happens here if there are new results uploaded since data was read in firebase? Are they not impacted by this `update`?

laurentS replied:
I'm thinking it would probably be useful to write a unit test for this:
- create fake result data in firebase
- read it with the function above
- simulate a user uploading a result for a different group
- delete_results_from_firebase() and check that the extra result has not been removed/altered

Hagellach37 replied:
Hey,
thanks for this comment. I added a new test for this case and check that the results that are uploaded only after the transfer results process has started, are not deleted.

laurentS reviewed (changes requested):
I think there's a couple of change to make to help limit data loss. I don't know enough about how firebase works with updates, so maybe worth checking the docs a bit more (unless you know the answers to the points below from both Matthias and myself).
The key problem you're trying to solve here is write conflicts, when multiple users write to firebase at the same time, and that's hard to test for. The best I can think of would be to do a bit of load testing using something like https://locust.io/, and simulate N writes within a given time period, and check that each of them ends up in postgres.

Hagellach37 commented on mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py:
```diff
@@ -9,37 +9,24 @@
+    Will not transfer results for tutorials and
+    for projects which are not set up in postgres.
+    """
     if not project_id_list:
```
adjusted

Hagellach37 commented:
Hey @laurentS  @Matthias-Schaub 
thanks for your thoughtful comments. This is exactly what I was hoping for.

Based on your suggestions I adjusted the code and added new tests as well.

Maybe one of you can take a look again if this satisfies you.

Hagellach37 commented on mapswipe_workers/tests/integration/test_transfer_results.py:
```diff
@@ -81,6 +86,52 @@ def test_user_not_in_postgres(self):
         result = pg_db.retr_query(sql_query)
         self.assertIsNotNone(result)
 
+    def test_additional_results_are_not_deleted(self):
```
@laurentS this is the test that you suggested.

laurentS replied:
I don't know enough about how this works, apologies if I'm asking dumb stuff: the initial results are apparently created in group 115, so this test does not check for group clash, right? Would it be worth adding an extra test where you have additional results in the same group 115 during the transfer process? Basically the same as this one, only changing the groupId in the extra results.

Hagellach37 replied:
I added another test for this. So we have two test cases now:

1. a new result for a different group in the results
2. a new result for an existing group in the results

Hagellach37 commented on mapswipe_workers/tests/integration/test_transfer_results_invalid_results.py:
```diff
@@ -0,0 +1,51 @@
+    def tearDown(self):
+        tear_down.delete_test_data(self.project_id)
+
+    def test_invalid_task_in_result(self):
```
@laurentS here I test for an psycopg2 related error during the transfer of the results

laurentS commented on mapswipe_workers/tests/integration/test_transfer_results_invalid_results.py:
```diff
@@ -0,0 +1,51 @@
+
+        fb_db = auth.firebaseDB()
+        ref = fb_db.reference("v2/results/{0}".format(self.project_id))
+        self.assertDictEqual(ref.get(shallow=True), {"g115": True})
```
Maybe I am being paranoid, but this is not actually checking that the results are still there, just verifying the presence of the key. Would checking something like  `fb_db.reference("test_build_area/results/18-156540-142783").get == 1` make sense?

Hagellach37 replied:
Hey,
I adjusted this part and hope that it is now "paranoid save". :D

I check for the number of results, rather than for a specific value.

laurentS commented on mapswipe_workers/tests/integration/test_transfer_results_many_results.py:
```diff
@@ -0,0 +1,39 @@
+            f"SELECT * "
+            f"FROM results "
+            f"WHERE project_id = '{self.project_id}' "
+            f"AND user_id = '{self.project_id}'"
```
Is the user_id meant to be the same as the project_id?

Hagellach37 replied:
yes, it's a bit confusing but in this case they are really the same.

Hagellach37 commented:
Hey @laurentS ,
anything more that I should do here? Or can I merge and deploy to the dev server?

laurentS commented:
> anything more that I should do here? Or can I merge and deploy to the dev server?

Nothing I can think of. But we should probably try to load test the dev server at some point
